File: python/algorithms/sorting/ex3_merge_sort.py
"""
Exercise: Merge Sort (Medium)

Implement the merge sort algorithm to sort a list of integers in ascending order.
Merge sort is a divide-and-conquer algorithm that recursively divides the list into
two halves, sorts each half, and then merges them back together.

Example:
Input: [38, 27, 43, 3, 9, 82, 10]
Output: [3, 9, 10, 27, 38, 43, 82]

Example (already sorted):
Input: [1, 2, 3, 4, 5]
Output: [1, 2, 3, 4, 5]

Time Complexity: O(n log n) for all cases (best, average, worst)
Space Complexity: O(n) - requires auxiliary array for merging
Stability: Yes - maintains relative order of equal elements
"""

from typing import List


def merge_sort(arr: List[int]) -> None:
    """
    Sorts a list using the merge sort algorithm (in-place).

    Args:
        arr: List of integers to be sorted (modified in place)

    Returns:
        None (the list is sorted in place)
    """
    # TODO: Implement merge sort
    # Hint 1: Base case - if len(arr) <= 1, return
    # Hint 2: Find the middle point and split into left and right halves
    # Hint 3: Recursively call merge_sort on both halves
    # Hint 4: Merge the two sorted halves back into arr
    # Hint 5: You can use slicing or indices to work with subarrays
    raise NotImplementedError("Not implemented yet")


def _merge(arr: List[int], left: int, mid: int, right: int) -> None:
    """
    Helper function to merge two sorted subarrays arr[left..mid] and arr[mid+1..right].

    Args:
        arr: The list containing both subarrays
        left: Starting index of left subarray
        mid: Ending index of left subarray
        right: Ending index of right subarray
    """
    # TODO: Implement merge logic
    # Hint 1: Create temporary lists for left and right halves
    # Hint 2: Use three pointers to merge the temporary lists back into arr
    raise NotImplementedError("Not implemented yet")


if __name__ == "__main__":
    # Example usage
    test_arr = [38, 27, 43, 3, 9, 82, 10]
    print(f"Before sorting: {test_arr}")
    merge_sort(test_arr)
    print(f"After sorting: {test_arr}")

File: rust/algorithms/src/sorting/ex3_merge_sort.rs
/// Exercise: Merge Sort (Medium)
///
/// Implement the merge sort algorithm to sort a vector of integers in ascending order.
/// Merge sort is a divide-and-conquer algorithm that recursively divides the vector into
/// two halves, sorts each half, and then merges them back together.
///
/// Example:
/// Input: &mut [38, 27, 43, 3, 9, 82, 10]
/// Output: [3, 9, 10, 27, 38, 43, 82]
///
/// Example (already sorted):
/// Input: &mut [1, 2, 3, 4, 5]
/// Output: [1, 2, 3, 4, 5]
///
/// Time Complexity: O(n log n) for all cases (best, average, worst)
/// Space Complexity: O(n) - requires auxiliary array for merging
/// Stability: Yes - maintains relative order of equal elements

pub fn merge_sort(arr: &mut [i32]) {
    // TODO: Implement merge sort
    // Hint 1: Base case - if arr.len() <= 1, return
    // Hint 2: Find the middle point and split into left and right halves
    // Hint 3: Recursively call merge_sort on both halves
    // Hint 4: Merge the two sorted halves back
    // Hint 5: You may need a helper function or use Vec for temporary storage
    unimplemented!("Not implemented yet")
}

/// Helper function to merge two sorted slices into a single sorted array
fn merge(left: &[i32], right: &[i32], result: &mut [i32]) {
    // TODO: Implement merge logic
    // Hint 1: Use two pointers, one for left and one for right
    // Hint 2: Compare elements and copy the smaller one to result
    // Hint 3: Copy any remaining elements from either left or right
    unimplemented!("Not implemented yet")
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_basic_case() {
        let mut arr = vec![38, 27, 43, 3, 9, 82, 10];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![3, 9, 10, 27, 38, 43, 82]);
    }

    #[test]
    fn test_already_sorted() {
        let mut arr = vec![1, 2, 3, 4, 5];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_reverse_sorted() {
        let mut arr = vec![5, 4, 3, 2, 1];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![1, 2, 3, 4, 5]);
    }

    #[test]
    fn test_single_element() {
        let mut arr = vec![42];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![42]);
    }

    #[test]
    fn test_empty_array() {
        let mut arr: Vec<i32> = vec![];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![]);
    }

    #[test]
    fn test_duplicates() {
        let mut arr = vec![3, 1, 4, 1, 5, 9, 2, 6, 5];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![1, 1, 2, 3, 4, 5, 5, 6, 9]);
    }

    #[test]
    fn test_all_same() {
        let mut arr = vec![5, 5, 5, 5, 5];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![5, 5, 5, 5, 5]);
    }

    #[test]
    fn test_negative_numbers() {
        let mut arr = vec![-5, 2, -8, 1, 9];
        merge_sort(&mut arr);
        assert_eq!(arr, vec![-8, -5, 1, 2, 9]);
    }
}
